### ISSUE TITLE
Long strings are truncated on export

### ISSUE THREAD

**Issue by etoledom** [GlotPress/GlotPress-WP#921]
Reference: [This ticket](https://github.com/wordpress-mobile/WordPress-iOS/issues/10006) in the iOS WordPress App.

We have an issue with a long string:

> "If you want a site but don't want any of the posts and pages you have now, our support team can delete your posts, pages, media, and comments for you.\n\nThis will keep your site and URL active, but give you a fresh start on your content creation. Just contact us to have your current content cleared out."

It uploads correctly. when we retrieve the translations, the translated part is correct but the key part (at the left side) is truncated:

>"If you want a site but don't want any of the posts and pages you have now, our support team can delete your posts, pages, media, and comments for you.\n\nThis will keep your site and URL active, but give you a fresh start on your content creation. Just cont"

You can see it exporting this file with this url:
[https://translate.wordpress.org/projects/apps/ios/dev/fr/default/export-translations?format=strings](https://translate.wordpress.org/projects/apps/ios/dev/fr/default/export-translations?format=strings)

The example is with the French translations but it happens with all the languages.

**Comment by toolstack** [comment]
Yep, that's because the  export code uses the context key, which is limited to 255 characters, instead of the original string, which isn't.

Should be an easy fix.

**Comment by AliSoftware** [comment]
Hello 👋 
I think we need to reopen this issue, as despite having been fixed by #922… it seems it got reintroduced by #1105 – which reverted the use of the context as key for the export.

I indeed recently found out that this key (the one in this issue description) created the same issue again in GP (we use `3.0.0-alpha.2`) and was truncated on export. For this time around, we ended up splitting the key in 2 smaller texts/keys instead but this can still be an issue for future keys.

**Comment by akirk** [comment]
@AliSoftware I understand that you feel this re-introduced a bug but I'd suggest to look at the unit test data in https://github.com/GlotPress/GlotPress-WP/blob/develop/tests/phpunit/data/translation.strings#L12-L16 which shows how #1105 is necessary to allow different translations for the same English singular text (the example was "Trash" as verb and as noun which need to be translated differently in many languages).

The limitation of a context to 255 chars is due to the context database field which is a `varchar(255)`. If you use your own GlotPress install and a database that can handle longer varchars, the easiest fix (without modifying your approach to translatable text in your iOS app, see below) for you would be to just increase the length of the field.

A context longer than 255 chars will be already truncated on import of the originals (and thus translators don't see the full context while translating already), so I think a better solution would be to add a warning upon import.

As a matter of fact, there is no need for a context to be longer than 255 characters in an iOS app. If it does, you might be using a side-effect of [`NSLocalizedString` which is defined as](https://developer.apple.com/documentation/foundation/nslocalizedstring):
```
NSLocalizedString(key, comment)
```
Further is this behavior defined when extracting the strings using the `genstrings` utility:

> The initial value for key in the strings file will be key.

Therefore it is convenient to use the English text as a string. But, it actually is a *key* (which is effectively what we call "context" above). So you can put any non-colliding string there, you just need to change the "translation" for English, i.e. in `en.lproj/Localizable.strings`.

So for example

```objc
NSLocalizedString(@"short-key", "Comment for the translator") 
```

would be exported by `genstrings` to `en.lproj/Localizable.strings` as:

```
/* Comment for the translator */
"short-key" = "short-key"
```

But since this is just a translatable file, you can easily change this to:
```
/* Comment for the translator */
"short-key" = "This is a very long text that exceeds 255 chars and would normally be truncated when importing into GlotPress which causes problems when exporting the translated texts since the originally specified context can no longer matched to the imported one, since it was truncated by the database."
```

This should solve your problem at the root.
